File: crates/lewm-hub/src/cost_ledger.rs
//! Cost-ledger parsing, appending, and integrity checks.

use std::{fmt, fs, path::Path, time::Duration};

/// Project hard cap for Hugging Face Jobs spend, in cents.
pub const HARD_CAP_USD_CENTS: u64 = 20_000;

const LEDGER_HEADER: &str = r"# `lewm-rs` cost ledger

> Updated automatically by `lewm-hub::cost_ledger::append_entry` at every job termination.
> Manual entries are forbidden; use `cost_ledger::backfill --from <job_url>` to import.

| Date (UTC)          | Phase | Job ID            | Hardware     | Wall   | Cost (USD) | Cumulative (USD) | Notes |
|---------------------|-------|-------------------|--------------|--------|-----------:|----------------:|-------|
";

/// Monetary amount represented as whole USD cents.
#[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
pub struct UsdAmount {
    cents: u64,
}

impl UsdAmount {
    /// Construct an amount from USD cents.
    pub const fn from_cents(cents: u64) -> Self {
        Self { cents }
    }

    /// Return the amount in USD cents.
    pub const fn cents(self) -> u64 {
        self.cents
    }

    /// Parse a non-negative amount with at most two decimal places.
    ///
    /// # Errors
    ///
    /// Returns an error when the amount is negative, malformed, or more precise
    /// than cents.
    pub fn parse(value: &str) -> Result<Self, CostLedgerError> {
        let trimmed = value.trim();
        if trimmed.is_empty() || trimmed.starts_with('-') || trimmed.starts_with('+') {
            return Err(CostLedgerError::InvalidAmount(value.to_owned()));
        }

        let (dollars_part, cents_part) = match trimmed.split_once('.') {
            Some((dollars, cents)) => (dollars, Some(cents)),
            None => (trimmed, None),
        };

        if dollars_part.is_empty()
            || !dollars_part
                .chars()
                .all(|character| character.is_ascii_digit())
        {
            return Err(CostLedgerError::InvalidAmount(value.to_owned()));
        }

        let dollars = dollars_part
            .parse::<u64>()
            .map_err(|_| CostLedgerError::InvalidAmount(value.to_owned()))?;
        let dollars_cents = dollars
            .checked_mul(100)
            .ok_or(CostLedgerError::AmountOverflow)?;
        let cents = match cents_part {
            None => 0,
            Some(part)
                if !part.is_empty()
                    && part.len() <= 2
                    && part.chars().all(|character| character.is_ascii_digit()) =>
            {
                let mut padded = part.to_owned();
                while padded.len() < 2 {
                    padded.push('0');
                }
                padded
                    .parse::<u64>()
                    .map_err(|_| CostLedgerError::InvalidAmount(value.to_owned()))?
            },
            Some(_) => return Err(CostLedgerError::InvalidAmount(value.to_owned())),
        };

        Ok(Self::from_cents(
            dollars_cents
                .checked_add(cents)
                .ok_or(CostLedgerError::AmountOverflow)?,
        ))
    }

    fn checked_add(self, other: Self) -> Result<Self, CostLedgerError> {
        Ok(Self::from_cents(
            self.cents
                .checked_add(other.cents)
                .ok_or(CostLedgerError::AmountOverflow)?,
        ))
    }
}

impl fmt::Display for UsdAmount {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "{}.{:02}", self.cents / 100, self.cents % 100)
    }
}

/// A ledger entry before the cumulative column is recomputed.
#[derive(Debug, Clone, Eq, PartialEq)]
pub struct CostEntry {
    /// UTC timestamp in `YYYY-MM-DD HH:MM:SS` form.
    pub date_utc: String,
    /// Project phase, for example `P3`.
    pub phase: String,
    /// Hugging Face Jobs id.
    pub job_id: String,
    /// Hugging Face hardware flavor.
    pub hardware: String,
    /// Wall-clock duration, for example `0:28:13`.
    pub wall: String,
    /// Conservatively rounded cost in USD.
    pub cost_usd: UsdAmount,
    /// Free-form note without Markdown table delimiters.
    pub notes: String,
}

/// A ledger row with its stored cumulative amount.
#[derive(Debug, Clone, Eq, PartialEq)]
pub struct CostLedgerRow {
    /// Entry fields for the row.
    pub entry: CostEntry,
    /// Stored cumulative amount for this row.
    pub cumulative_usd: UsdAmount,
}

/// Cost-ledger failures.
#[derive(Debug, thiserror::Error, Eq, PartialEq)]
pub enum CostLedgerError {
    /// Filesystem operation failed.
    #[error("cost-ledger I/O failed: {0}")]
    Io(String),

    /// Markdown ledger content did not match the expected table format.
    #[error("invalid cost-ledger content: {0}")]
    InvalidLedger(String),

    /// A money value could not be parsed as USD cents.
    #[error("invalid USD amount: {0}")]
    InvalidAmount(String),

    /// A field cannot be rendered safely inside a Markdown table cell.
    #[error("invalid Markdown table field {field}: {value}")]
    InvalidField {
        /// Field name.
        field: &'static str,
        /// Invalid value.
        value: String,
    },

    /// Cumulative spend exceeded the configured cap.
    #[error("cost-ledger cap exceeded: cumulative {cumulative} USD > cap {cap} USD")]
    CapExceeded {
        /// Cumulative amount.
        cumulative: UsdAmount,
        /// Configured cap.
        cap: UsdAmount,
    },

    /// Stored cumulative column does not match the recomputed cumulative sum.
    #[error(
        "cost-ledger cumulative mismatch at row {row}: stored {stored} USD, expected {expected} USD"
    )]
    CumulativeMismatch {
        /// One-based ledger row number.
        row: usize,
        /// Stored cumulative amount.
        stored: UsdAmount,
        /// Recomputed cumulative amount.
        expected: UsdAmount,
    },

    /// Integer overflow while computing monetary amounts.
    #[error("cost-ledger amount overflow")]
    AmountOverflow,
}

impl From<std::io::Error> for CostLedgerError {
    fn from(error: std::io::Error) -> Self {
        Self::Io(error.to_string())
    }
}

/// Append an entry, rewrite cumulative values from scratch, and enforce the hard cap.
///
/// Missing ledger files are created with the RFC 0010 header.
///
/// # Errors
///
/// Returns an error when the existing ledger is malformed, an entry cannot be
/// rendered safely, filesystem access fails, cumulative values are inconsistent,
/// or the hard cap would be exceeded.
pub fn append_entry(
    entry: CostEntry,
    ledger_path: &Path,
) -> Result<Vec<CostLedgerRow>, CostLedgerError> {
    let mut entries = read_ledger(ledger_path)?
        .into_iter()
        .map(|row| row.entry)
        .collect::<Vec<_>>();
    entries.push(entry);

    let rows = rows_from_entries(&entries, UsdAmount::from_cents(HARD_CAP_USD_CENTS))?;
    write_ledger(ledger_path, &rows)?;
    Ok(rows)
}

/// Read a ledger and verify its cumulative column against the hard cap.
///
/// # Errors
///
/// Returns an error when the ledger cannot be read or parsed, when cumulative
/// values are inconsistent, or when the hard cap is exceeded.
pub fn read_ledger(ledger_path: &Path) -> Result<Vec<CostLedgerRow>, CostLedgerError> {
    if !ledger_path.exists() {
        return Ok(Vec::new());
    }

    let raw = fs::read_to_string(ledger_path)?;
    let mut rows = Vec::new();
    for line in raw.lines() {
        if let Some(row) = parse_table_row(line)? {
            rows.push(row);
        }
    }
    verify_rows(&rows, UsdAmount::from_cents(HARD_CAP_USD_CENTS))?;
    Ok(rows)
}

/// Verify that a ledger is internally consistent and under a supplied cap.
///
/// # Errors
///
/// Returns an error when parsing fails, cumulative values are inconsistent, or
/// any cumulative row exceeds `cap`.
pub fn verify_ledger(
    ledger_path: &Path,
    cap: UsdAmount,
) -> Result<Vec<CostLedgerRow>, CostLedgerError> {
    if !ledger_path.exists() {
        return Ok(Vec::new());
    }

    let raw = fs::read_to_string(ledger_path)?;
    let mut rows = Vec::new();
    for line in raw.lines() {
        if let Some(row) = parse_table_row(line)? {
            rows.push(row);
        }
    }
    verify_rows(&rows, cap)?;
    Ok(rows)
}

/// Return the conservatively rounded billable minutes for a wall-clock duration.
pub fn rounded_billable_minutes(wall: Duration) -> u64 {
    let seconds = wall
        .as_secs()
        .saturating_add(u64::from(wall.subsec_nanos() > 0));
    seconds.div_ceil(60)
}

/// Compute cost from a wall-clock duration and a hardware price in cents/hour.
///
/// The wall time is rounded up to the nearest minute, then the final cost is
/// rounded up to the nearest cent so the ledger never understates spend.
///
/// # Errors
///
/// Returns an error on integer overflow.
pub fn cost_for_wall_time(
    wall: Duration,
    price_cents_per_hour: u64,
) -> Result<UsdAmount, CostLedgerError> {
    let minutes = rounded_billable_minutes(wall);
    let numerator = price_cents_per_hour
        .checked_mul(minutes)
        .ok_or(CostLedgerError::AmountOverflow)?;
    Ok(UsdAmount::from_cents(numerator.div_ceil(60)))
}

fn rows_from_entries(
    entries: &[CostEntry],
    cap: UsdAmount,
) -> Result<Vec<CostLedgerRow>, CostLedgerError> {
    let mut cumulative = UsdAmount::from_cents(0);
    let mut rows = Vec::with_capacity(entries.len());

    for entry in entries {
        validate_entry(entry)?;
        cumulative = cumulative.checked_add(entry.cost_usd)?;
        if cumulative > cap {
            return Err(CostLedgerError::CapExceeded { cumulative, cap });
        }
        rows.push(CostLedgerRow {
            entry: entry.clone(),
            cumulative_usd: cumulative,
        });
    }

    Ok(rows)
}

fn verify_rows(rows: &[CostLedgerRow], cap: UsdAmount) -> Result<(), CostLedgerError> {
    let mut expected = UsdAmount::from_cents(0);
    for (index, row) in rows.iter().enumerate() {
        validate_entry(&row.entry)?;
        expected = expected.checked_add(row.entry.cost_usd)?;
        if row.cumulative_usd != expected {
            return Err(CostLedgerError::CumulativeMismatch {
                row: index + 1,
                stored: row.cumulative_usd,
                expected,
            });
        }
        if row.cumulative_usd > cap {
            return Err(CostLedgerError::CapExceeded {
                cumulative: row.cumulative_usd,
                cap,
            });
        }
    }
    Ok(())
}

fn write_ledger(ledger_path: &Path, rows: &[CostLedgerRow]) -> Result<(), CostLedgerError> {
    if let Some(parent) = ledger_path.parent() {
        fs::create_dir_all(parent)?;
    }

    let mut rendered = LEDGER_HEADER.to_owned();
    for row in rows {
        rendered.push_str(&format_row(row)?);
    }
    fs::write(ledger_path, rendered)?;
    Ok(())
}

fn format_row(row: &CostLedgerRow) -> Result<String, CostLedgerError> {
    Ok(format!(
        "| {} | {} | {} | {} | {} | {} | {} | {} |\n",
        markdown_cell("date_utc", &row.entry.date_utc, false)?,
        markdown_cell("phase", &row.entry.phase, false)?,
        markdown_cell("job_id", &row.entry.job_id, false)?,
        markdown_cell("hardware", &row.entry.hardware, false)?,
        markdown_cell("wall", &row.entry.wall, false)?,
        row.entry.cost_usd,
        row.cumulative_usd,
        markdown_cell("notes", &row.entry.notes, true)?,
    ))
}

fn validate_entry(entry: &CostEntry) -> Result<(), CostLedgerError> {
    markdown_cell("date_utc", &entry.date_utc, false)?;
    markdown_cell("phase", &entry.phase, false)?;
    markdown_cell("job_id", &entry.job_id, false)?;
    markdown_cell("hardware", &entry.hardware, false)?;
    markdown_cell("wall", &entry.wall, false)?;
    markdown_cell("notes", &entry.notes, true)?;
    Ok(())
}

fn markdown_cell(
    field: &'static str,
    value: &str,
    allow_empty: bool,
) -> Result<String, CostLedgerError> {
    let trimmed = value.trim();
    if (!allow_empty && trimmed.is_empty())
        || trimmed.contains('|')
        || trimmed.contains('\n')
        || trimmed.contains('\r')
    {
        return Err(CostLedgerError::InvalidField {
            field,
            value: value.to_owned(),
        });
    }
    Ok(trimmed.to_owned())
}

fn parse_table_row(line: &str) -> Result<Option<CostLedgerRow>, CostLedgerError> {
    let trimmed = line.trim();
    if !trimmed.starts_with('|') || !trimmed.ends_with('|') {
        return Ok(None);
    }

    let cells = trimmed
        .trim_matches('|')
        .split('|')
        .map(str::trim)
        .collect::<Vec<_>>();

    match cells.as_slice() {
        [
            "Date (UTC)",
            "Phase",
            "Job ID",
            "Hardware",
            "Wall",
            "Cost (USD)",
            "Cumulative (USD)",
            "Notes",
        ] => Ok(None),
        [separator, _, _, _, _, _, _, _] if separator.chars().all(|character| character == '-') => {
            Ok(None)
        },
        [
            date_utc,
            phase,
            job_id,
            hardware,
            wall,
            cost_usd,
            cumulative_usd,
            notes,
        ] => {
            let entry = CostEntry {
                date_utc: (*date_utc).to_owned(),
                phase: (*phase).to_owned(),
                job_id: (*job_id).to_owned(),
                hardware: (*hardware).to_owned(),
                wall: (*wall).to_owned(),
                cost_usd: UsdAmount::parse(cost_usd)?,
                notes: (*notes).to_owned(),
            };
            let cumulative_usd = UsdAmount::parse(cumulative_usd)?;
            Ok(Some(CostLedgerRow {
                entry,
                cumulative_usd,
            }))
        },
        _ => Err(CostLedgerError::InvalidLedger(format!(
            "expected 8 table cells, got {} in line: {line}",
            cells.len()
        ))),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::{
        path::PathBuf,
        time::{SystemTime, UNIX_EPOCH},
    };

    #[test]
    fn cost_ledger_cumulative_correct() -> Result<(), Box<dyn std::error::Error>> {
        let path = unique_ledger_path("cumulative");

        append_entry(sample_entry("hfjob-1", 38), &path)?;
        let rows = append_entry(sample_entry("hfjob-2", 112), &path)?;
        let raw = fs::read_to_string(&path)?;

        assert_eq!(rows.len(), 2);
        assert!(raw.contains("| 2026-05-12 15:03:00 | P3 | hfjob-2 | a10g-large | 0:44:10 | 1.12 | 1.50 | smoke pusht |"));
        assert_eq!(verify_ledger(&path, UsdAmount::from_cents(20_000))?, rows);

        remove_file(&path)?;
        Ok(())
    }

    #[test]
    fn cost_ledger_under_200_usd() -> Result<(), Box<dyn std::error::Error>> {
        let path = unique_ledger_path("under-cap");

        append_entry(sample_entry("hfjob-1", 19_999), &path)?;
        let Err(error) = append_entry(sample_entry("hfjob-2", 2), &path) else {
            return Err("cap should be enforced".into());
        };

        assert_eq!(
            error,
            CostLedgerError::CapExceeded {
                cumulative: UsdAmount::from_cents(20_001),
                cap: UsdAmount::from_cents(20_000),
            }
        );

        remove_file(&path)?;
        Ok(())
    }

    #[test]
    fn cost_calculation_rounds_up_to_nearest_minute() -> Result<(), Box<dyn std::error::Error>> {
        assert_eq!(rounded_billable_minutes(Duration::from_secs(60)), 1);
        assert_eq!(rounded_billable_minutes(Duration::from_secs(61)), 2);
        assert_eq!(
            cost_for_wall_time(Duration::from_secs(60), 150)?,
            UsdAmount::from_cents(3)
        );
        assert_eq!(
            cost_for_wall_time(Duration::from_secs(61), 150)?,
            UsdAmount::from_cents(5)
        );
        Ok(())
    }

    fn sample_entry(job_id: &str, cost_cents: u64) -> CostEntry {
        CostEntry {
            date_utc: "2026-05-12 15:03:00".to_owned(),
            phase: "P3".to_owned(),
            job_id: job_id.to_owned(),
            hardware: "a10g-large".to_owned(),
            wall: "0:44:10".to_owned(),
            cost_usd: UsdAmount::from_cents(cost_cents),
            notes: "smoke pusht".to_owned(),
        }
    }

    fn unique_ledger_path(name: &str) -> PathBuf {
        let nanos = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map_or(0, |duration| duration.as_nanos());
        std::env::temp_dir().join(format!(
            "lewm-hub-cost-ledger-{name}-{}-{nanos}.md",
            std::process::id()
        ))
    }

    fn remove_file(path: &Path) -> Result<(), Box<dyn std::error::Error>> {
        if path.exists() {
            fs::remove_file(path)?;
        }
        Ok(())
    }
}

File: python/cost_ledger.py
"""Maintain reports/cost.md for Hugging Face Jobs spend."""

from __future__ import annotations

import argparse
import json
import subprocess
from collections.abc import Iterable, Mapping, Sequence
from dataclasses import dataclass
from decimal import Decimal
from pathlib import Path

from hf_pricing import estimate_cost_usd, format_usd, format_wall, parse_timestamp

DEFAULT_LEDGER = Path("reports/cost.md")
DEFAULT_CAP_USD = Decimal("200.00")
HEADER = """# `lewm-rs` cost ledger

> Updated automatically by `lewm-hub::cost_ledger::append_entry` at every job termination.
> Manual entries are forbidden; use `cost_ledger::backfill --from <job_url>` to import.

| Date (UTC)          | Phase | Job ID            | Hardware     | Wall   | Cost (USD) | Cumulative (USD) | Notes |
|---------------------|-------|-------------------|--------------|--------|-----------:|----------------:|-------|
"""


@dataclass(frozen=True)
class LedgerRow:
    """One cost-ledger row."""

    date_utc: str
    phase: str
    job_id: str
    hardware: str
    wall: str
    cost_usd: Decimal
    cumulative_usd: Decimal
    notes: str


def parse_usd(value: str) -> Decimal:
    """Parse a USD amount at cent precision."""
    amount = Decimal(value.strip()).quantize(Decimal("0.01"))
    if amount < 0:
        raise ValueError(f"negative USD amount: {value}")
    return amount


def read_ledger(path: Path) -> list[LedgerRow]:
    """Read a Markdown cost ledger."""
    if not path.exists():
        return []

    rows: list[LedgerRow] = []
    for line in path.read_text(encoding="utf-8").splitlines():
        stripped = line.strip()
        if not stripped.startswith("|") or not stripped.endswith("|"):
            continue
        cells = [cell.strip() for cell in stripped.strip("|").split("|")]
        if cells == [
            "Date (UTC)",
            "Phase",
            "Job ID",
            "Hardware",
            "Wall",
            "Cost (USD)",
            "Cumulative (USD)",
            "Notes",
        ]:
            continue
        if len(cells) == 8 and set(cells[0]) == {"-"}:
            continue
        if len(cells) != 8:
            raise ValueError(f"expected 8 cells in ledger row: {line}")
        rows.append(
            LedgerRow(
                date_utc=cells[0],
                phase=cells[1],
                job_id=cells[2],
                hardware=cells[3],
                wall=cells[4],
                cost_usd=parse_usd(cells[5]),
                cumulative_usd=parse_usd(cells[6]),
                notes=cells[7],
            )
        )
    verify_rows(rows, DEFAULT_CAP_USD)
    return rows


def verify_rows(rows: Sequence[LedgerRow], cap_usd: Decimal) -> None:
    """Verify cumulative values and the configured spending cap."""
    cumulative = Decimal("0.00")
    for index, row in enumerate(rows, start=1):
        cumulative += row.cost_usd
        cumulative = cumulative.quantize(Decimal("0.01"))
        if row.cumulative_usd != cumulative:
            raise ValueError(
                f"row {index} cumulative mismatch: stored {row.cumulative_usd}, expected {cumulative}"
            )
        if row.cumulative_usd > cap_usd:
            raise ValueError(f"cost cap exceeded at row {index}: {row.cumulative_usd} > {cap_usd}")


def write_ledger(path: Path, rows: Iterable[LedgerRow]) -> None:
    """Write a Markdown cost ledger with recomputed cumulative values."""
    path.parent.mkdir(parents=True, exist_ok=True)
    cumulative = Decimal("0.00")
    rendered = HEADER
    for row in rows:
        cumulative += row.cost_usd
        cumulative = cumulative.quantize(Decimal("0.01"))
        rendered += (
            f"| {cell(row.date_utc, 'date_utc')} | {cell(row.phase, 'phase')} | "
            f"{cell(row.job_id, 'job_id')} | {cell(row.hardware, 'hardware')} | "
            f"{cell(row.wall, 'wall')} | {format_usd(row.cost_usd)} | "
            f"{format_usd(cumulative)} | {cell(row.notes, 'notes', allow_empty=True)} |\n"
        )
    path.write_text(rendered, encoding="utf-8")


def cell(value: str, field: str, *, allow_empty: bool = False) -> str:
    """Validate and normalize a Markdown table cell."""
    normalized = value.strip()
    if (not allow_empty and not normalized) or "|" in normalized or "\n" in normalized or "\r" in normalized:
        raise ValueError(f"invalid ledger field {field}: {value!r}")
    return normalized


def append_row(path: Path, row: LedgerRow, cap_usd: Decimal) -> list[LedgerRow]:
    """Append a row, recompute cumulative values, and write the ledger."""
    existing = read_ledger(path)
    rows = [*existing, row_without_cumulative(row)]
    recomputed = recompute(rows, cap_usd)
    write_ledger(path, recomputed)
    return recomputed


def row_without_cumulative(row: LedgerRow) -> LedgerRow:
    """Drop a stale cumulative value before recomputing."""
    return LedgerRow(
        date_utc=row.date_utc,
        phase=row.phase,
        job_id=row.job_id,
        hardware=row.hardware,
        wall=row.wall,
        cost_usd=row.cost_usd,
        cumulative_usd=Decimal("0.00"),
        notes=row.notes,
    )


def recompute(rows: Sequence[LedgerRow], cap_usd: Decimal) -> list[LedgerRow]:
    """Recompute cumulative values for all rows."""
    cumulative = Decimal("0.00")
    recomputed: list[LedgerRow] = []
    for index, row in enumerate(rows, start=1):
        cumulative += row.cost_usd
        cumulative = cumulative.quantize(Decimal("0.01"))
        if cumulative > cap_usd:
            raise ValueError(f"cost cap exceeded at row {index}: {cumulative} > {cap_usd}")
        recomputed.append(
            LedgerRow(
                date_utc=row.date_utc,
                phase=row.phase,
                job_id=row.job_id,
                hardware=row.hardware,
                wall=row.wall,
                cost_usd=row.cost_usd,
                cumulative_usd=cumulative,
                notes=row.notes,
            )
        )
    return recomputed


def row_from_job_record(record: Mapping[str, object]) -> LedgerRow:
    """Convert an HF Jobs record to a ledger row."""
    job_id = str(record.get("job_id") or record.get("id") or "").strip()
    hardware = str(record.get("hardware_flavor") or record.get("hardware") or "").strip()
    started_at = str(record.get("started_at") or record.get("startedAt") or "").strip()
    ended_at = str(record.get("ended_at") or record.get("endedAt") or "").strip()
    if not job_id or not hardware or not started_at or not ended_at:
        raise ValueError(f"job record is missing required fields: {record}")
    phase = str(record.get("phase") or "unknown").strip()
    exit_code = record.get("exit_code", record.get("exitCode", "unknown"))
    notes = str(record.get("notes") or f"exit_code={exit_code}").strip()
    return LedgerRow(
        date_utc=parse_timestamp(ended_at).strftime("%Y-%m-%d %H:%M:%S"),
        phase=phase,
        job_id=job_id,
        hardware=hardware,
        wall=format_wall(started_at, ended_at),
        cost_usd=estimate_cost_usd(hardware, started_at, ended_at),
        cumulative_usd=Decimal("0.00"),
        notes=notes,
    )


def load_job_records(
    input_json: Path | None,
    since: str | None,
    org: str,
    from_url: str | None,
) -> list[Mapping[str, object]]:
    """Load HF Jobs records from JSON or by shelling out to `hf jobs list`."""
    if input_json is not None:
        payload = json.loads(input_json.read_text(encoding="utf-8"))
    elif from_url is not None:
        command = ["hf", "jobs", "inspect", from_url, "--json"]
        completed = subprocess.run(command, check=True, text=True, capture_output=True)
        payload = json.loads(completed.stdout)
    else:
        if since is None:
            raise ValueError("backfill requires --since, --from, or --input-json")
        command = ["hf", "jobs", "list", "--org", org, "--since", since, "--json"]
        completed = subprocess.run(command, check=True, text=True, capture_output=True)
        payload = json.loads(completed.stdout)
    if isinstance(payload, dict):
        if "jobs" in payload:
            payload = payload["jobs"]
        else:
            payload = [payload]
    if not isinstance(payload, list):
        raise ValueError("HF Jobs payload must be a list or an object with a jobs list")
    return [record for record in payload if isinstance(record, Mapping)]


def check_command(args: argparse.Namespace) -> None:
    """Run the ledger integrity and cap check."""
    rows = read_ledger(args.path)
    verify_rows(rows, Decimal(args.cap_usd).quantize(Decimal("0.01")))
    print(f"cost ledger ok: {len(rows)} rows, cap {Decimal(args.cap_usd).quantize(Decimal('0.01'))} USD")


def append_command(args: argparse.Namespace) -> None:
    """Append one job row."""
    if args.cost_usd is not None:
        if args.wall is None:
            raise ValueError("--wall is required with --cost-usd")
        cost_usd = parse_usd(args.cost_usd)
        wall = args.wall
    else:
        if args.started_at is None or args.ended_at is None:
            raise ValueError("--started-at and --ended-at are required when --cost-usd is absent")
        cost_usd = estimate_cost_usd(args.hardware, args.started_at, args.ended_at)
        wall = format_wall(args.started_at, args.ended_at)

    row = LedgerRow(
        date_utc=args.date_utc,
        phase=args.phase,
        job_id=args.job_id,
        hardware=args.hardware,
        wall=wall,
        cost_usd=cost_usd,
        cumulative_usd=Decimal("0.00"),
        notes=args.notes,
    )
    rows = append_row(args.path, row, Decimal(args.cap_usd).quantize(Decimal("0.01")))
    print(f"appended {args.job_id}; rows={len(rows)}")


def backfill_command(args: argparse.Namespace) -> None:
    """Backfill missing jobs into the ledger."""
    rows = read_ledger(args.path)
    known_job_ids = {row.job_id for row in rows}
    new_rows = [row_without_cumulative(row) for row in rows]
    appended = 0
    for record in load_job_records(args.input_json, args.since, args.org, args.from_url):
        row = row_from_job_record(record)
        if row.job_id in known_job_ids:
            continue
        new_rows.append(row)
        known_job_ids.add(row.job_id)
        appended += 1
    recomputed = recompute(new_rows, Decimal(args.cap_usd).quantize(Decimal("0.01")))
    write_ledger(args.path, recomputed)
    print(f"backfilled {appended} jobs; rows={len(recomputed)}")


def build_parser() -> argparse.ArgumentParser:
    """Build the command-line parser."""
    parser = argparse.ArgumentParser(description=__doc__)
    subcommands = parser.add_subparsers(dest="command", required=True)

    check = subcommands.add_parser("check", help="verify cumulative values and cap")
    check.add_argument("--path", type=Path, default=DEFAULT_LEDGER)
    check.add_argument("--cap-usd", default=str(DEFAULT_CAP_USD))
    check.set_defaults(func=check_command)

    append = subcommands.add_parser("append", help="append one ledger row")
    append.add_argument("--path", type=Path, default=DEFAULT_LEDGER)
    append.add_argument("--cap-usd", default=str(DEFAULT_CAP_USD))
    append.add_argument("--date-utc", required=True)
    append.add_argument("--phase", required=True)
    append.add_argument("--job-id", required=True)
    append.add_argument("--hardware", required=True)
    append.add_argument("--started-at")
    append.add_argument("--ended-at")
    append.add_argument("--wall")
    append.add_argument("--cost-usd")
    append.add_argument("--notes", default="")
    append.set_defaults(func=append_command)

    backfill = subcommands.add_parser("backfill", help="backfill missing HF Jobs rows")
    backfill.add_argument("--path", type=Path, default=DEFAULT_LEDGER)
    backfill.add_argument("--cap-usd", default=str(DEFAULT_CAP_USD))
    backfill.add_argument("--since")
    backfill.add_argument("--from", dest="from_url")
    backfill.add_argument("--org", default="AbdelStark")
    backfill.add_argument("--input-json", type=Path)
    backfill.set_defaults(func=backfill_command)

    return parser


def main(argv: Sequence[str] | None = None) -> int:
    """Run the CLI."""
    args = build_parser().parse_args(argv)
    args.func(args)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
